cache/only_one_test.go

```diff
@@ -87,3 +88,46 @@ func TestOnlyOne_ComputeConcurrentlyOnce(t *testing.T) {
 		t.Error("did not expect to run concurrent compute after 5ms")
 	}
 }
+
+func TestOnlyOne_ComputeConcurrentlyOnceKeyedByStruct(t *testing.T) {
+	c := cache.NewChanOnlyOne()
+
+	type Key struct {
+		a, b int
+	}
+	const (
+		size        = 3
+		repeats     = 50
+		parallelism = 9
+	)
+
+	for i := 0; i < repeats; i++ {
+		var wg sync.WaitGroup
+		wg.Add(size * size)
+		ch := make(chan struct{})
+		numExec := int32(0)
+		for j := 0; j < parallelism; j++ {
+			go func() {
+				<-ch
+				for k := 0; k < size; k++ {
+					for l := 0; l < size; l++ {
+						v, err := c.Compute(Key{k, l}, func() (interface{}, error) {
+							atomic.AddInt32(&numExec, 1)
+							return Key{k + 1, l + 1}, nil
+						})
+						if err != nil {
+							t.Error(err)
+							continue
+						}
+						got := v.(Key)
+						expected := Key{a: k + 1, b: l + 1}
```

Comment:
Even if `Compute` is completely broken, that will always succeed, no?

You increase the incoming values by 1, even if they run in parallel for the same key, than you will still get the same output.